tests/invalid_models_tests/test_models.py

```diff
@@ -495,6 +495,34 @@ class Meta:
 
         self.assertEqual(Model.check(databases=self.databases), [])
 
+    def test_index_with_expressions(self):
+        class Model(models.Model):
+            name = models.TextField()
+
+            class Meta:
+                indexes = [
+                    models.Index(
+                        Upper('name'),
+                        name='index_upper_name',
+                    ),
+                ]
+
+        errors = Model.check(databases=self.databases)
+        expected = (
+            []
+            if connection.features.supports_expression_indexes
+            else [
+                Warning(
+                    '%s does not support indexes on expressions.'
+                    % connection.display_name,
+                    hint="An index won't be created. Silence this warning if you don't care about it.",
+                    obj=Model,
+                    id='models.W041',
```

Comment:
```suggestion
                    id='models.W042',
```
This should be documented (`docs/ref/checks.txt`).